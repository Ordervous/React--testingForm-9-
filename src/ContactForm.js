// Your form should meet the following requirements:

//     The form should have three fields: name, email, and phone with an "Add Contact" button.
//     Clicking on the "Add Contact" button invokes an onContactAdded callback function that is passed in via a prop.
//     It does not throw an error if the onContactAdded prop is undefined.

import React from "react";
import { Button, Form } from "react-bootstrap";

function OnContactAdded() {
    alert('Contact Added!')
}

function ContactForm(props) {


    return (
        <Form role="form">
            <Form.Group className="mb-3" data-testId="formBasicName">
                <Form.Label>Name</Form.Label>
                <Form.Control type="text" data-testId="inputname" placeholder="Enter name" />
            </Form.Group>

            <Form.Group className="mb-3" controlId="formBasicEmail">
                <Form.Label>Email address</Form.Label>
                <Form.Control type="email" placeholder="Enter Email" />
            </Form.Group>
            <Form.Group className="mb-3" controlId="formBasicPhone">
                <Form.Label>Phone Number</Form.Label>
                <Form.Control type="number" placeholder="Enter Phone" />
            </Form.Group>
            <Button variant="primary" type="submit" onClick={OnContactAdded}>Add Contact</Button>
        </Form>
    )
}
export default ContactForm

// snapshot test