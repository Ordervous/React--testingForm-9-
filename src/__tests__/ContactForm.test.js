// Your tests should at least verify the following:

//     All form fields are present on the page.
//     The button is present on the page.
//     Clicking the button invokes the callback. This should be verified using a mock function passed as a prop.
//     Entering text into a form field updates the value of the field.
//     The form renders without an onContactAdded prop.
//     If the markup of the page changes, a snapshot test will fail.

import React from "react";
import ContactForm from '../ContactForm';
import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event'


it('renders without error', () => {
    render(<ContactForm />)
})

afterEach(() => {
    jest.restoreAllMocks()
})

it('checks for a button on the page', () => {
    render(<ContactForm />)
    expect(screen.getByRole('button', {name: "Add Contact"})).toBeInTheDocument()
})

it('checks for a form on the page', () => {
    render(<ContactForm />)
    expect(screen.getByRole('form')).toBeInTheDocument()
})

it('displays the alert when the Add Contact button is clicked', () => {
    window.alert = jest.fn()
    render(<ContactForm />)
    const button = screen.getByRole('button', { name: "Add Contact" })
    fireEvent.click(button)
    expect(window.alert).toHaveBeenCalledTimes(1)
}) 

// Testing Input fields
it('test input Name field', () => {
    render(<ContactForm />)
    const inputname = screen.getByTestId("formBasicName")
    expect(inputname).toBeInTheDocument()
})

it('pass in text to Name input field', () => {
    render(<ContactForm />)
    const inputname = screen.getByTestId("inputname")
    userEvent.type(inputname, "Joe")

    expect(screen.getByTestId("inputname")).toHaveValue("Joe");
    expect(screen.queryByTestId("error-msg")).not.toBeInTheDocument();
})

it('pass in text to Name input field2', () => {
    render(<ContactForm />)
    const input = screen.getByTestId("inputname")
    fireEvent.change(input, { target: { value: 'John' } })
    expect(input.value).toBe('John')
})

it('if OnContactAdded props is undefined', () => {
    window.alert = jest.fn()
    render(<ContactForm />)
    const unprops = screen.queryByText('undefined')
    expect(unprops).not.toBeInTheDocument()
})
 

it('should match the snapshot', () => {
    const spy = jest.spyOn(Date, 'now')
      .mockImplementation(() => new Date('2021-01-01T00:00')
    )
    const { asFragment } = render(<ContactForm />)
    const html = asFragment()
    expect(html).toMatchSnapshot()
  })