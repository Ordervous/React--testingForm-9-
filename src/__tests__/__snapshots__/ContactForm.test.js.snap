// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<DocumentFragment>
  <form
    class=""
    role="form"
  >
    <div
      class="mb-3"
      data-testid="formBasicName"
    >
      <label
        class="form-label"
      >
        Name
      </label>
      <input
        class="form-control"
        data-testid="inputname"
        placeholder="Enter name"
        type="text"
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="formBasicEmail"
      >
        Email address
      </label>
      <input
        class="form-control"
        id="formBasicEmail"
        placeholder="Enter Email"
        type="email"
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="formBasicPhone"
      >
        Phone Number
      </label>
      <input
        class="form-control"
        id="formBasicPhone"
        placeholder="Enter Phone"
        type="number"
      />
    </div>
    <button
      class="btn btn-primary"
      type="submit"
    >
      Add Contact
    </button>
  </form>
</DocumentFragment>
`;
